nosnoc/problem.py

```diff
@@ -498,21 +517,30 @@ def forward_simulation(self, ocp: NosnocOcp, Uk: ca.SX, sot: ca.SX) -> None:
 
         for j in range(opts.n_s):
             # Dynamics excluding complementarities
-        
-            fj = sot * model.f_x_fun(X_fe[j], self.rk_stage_z(j), Uk, self.p, model.v_global)
             qj = sot * ocp.f_q_fun(X_fe[j], Uk, self.p, model.v_global)
-            gj = model.g_z_all_fun(X_fe[j], self.rk_stage_z(j), Uk, self.p)
-            self.add_constraint(gj)
             if opts.dcs_mode == DcsMode.PDS:
-                # Add PDS-specific constraints
+                # Get lambda for this stage
                 lam = self.w[flatten(self.ind_lam[j])]
+                # Pass lambda as additional argument
+                fj = sot * model.f_x_fun(X_fe[j], self.rk_stage_z(j), Uk, self.p, model.v_global, lam)
```

Comment:
If you correctly set up lambda as an algebraic variable you should be able to re-use the same call to `model.f_x_fun` between all dcs modes.